### PR TITLE
refactor(context): change context to pimpl

### PR DESCRIPTION
**Pull request by antmat**
Some refactor against context_t:
  * context_t is pimpl now
  * get rid of template methods or use explicit template instantiation
  * include optimization
This refactor gives about 10% boost in compilation time.
I have some more changes planned, but let's decide if we need to split it to several PR or not.

### CODE REVIEW

File: include/cocaine/context.hpp [hunk 1]
```diff
@@ -44,37 +35,15 @@ namespace signal {
...
-
-    // Context signalling hub.
-    retroactive_signal<io::context_tag> m_signals;
-
-public:
-    const config_t config;
-
-    // Service port mapping and pinning.
-    port_mapping_t mapper;
+    struct impl_t;
```
Comment (3Hren) [inline include/cocaine/context.hpp]: Maybe name it `context_data`, because it's only data, not implementation?

File: src/service/logging.cpp [hunk 2]
```diff
@@ -80,13 +82,13 @@ logging_t::logging_t(context_t& context, asio::io_service& asio, const std::stri
...
             std::stringstream stream;
-            stream << boost::lexical_cast<std::string>(context.config.logging.loggers);
+            stream << boost::lexical_cast<std::string>(context.config().logging.loggers);
 
             auto log = registry.builder<blackhole::config::json_t>(stream)
                 .build(backend);
 
             log.filter([&](const blackhole::record_t& record) -> bool {
-                return record.severity() >= context.config.logging.severity || !trace_t::current().empty();
+                return record.severity() >= context.config().logging.severity || !trace_t::current().empty();
```
Comment (3Hren) [inline src/service/logging.cpp]: Well, now it's not cheap as it was. Save severity.

File: src/context.cpp [hunk 3]
```diff
@@ -103,62 +248,63 @@ struct match {
...
         list.emplace_back(name, std::move(service));
     });
 
     // Fire off the signal to alert concerned subscribers about the service removal event.
-    m_signals.invoke<context::service::exposed>(actor.prototype().name(), std::forward_as_tuple(
-        actor.endpoints(),
-        actor.prototype().version(),
-        actor.prototype().root()
+    impl->signals.invoke<context::service::exposed>(actor.prototype().name(), std::forward_as_tuple(
+    actor.endpoints(),
```
Comment (3Hren) [inline src/context.cpp]: Strange indents.

File: src/context.cpp [hunk 4]
```diff
@@ -38,42 +42,168 @@
...
+    if (!errored.empty()) {
+        COCAINE_LOG_ERROR(impl->log, "emergency core shutdown");
+
+        // Signal and stop all the services, shut down execution units.
+        terminate();
+
+        std::ostringstream stream;
+        std::ostream_iterator<char> builder(stream);
+
+        boost::spirit::karma::generate(builder, boost::spirit::karma::string % ", ", errored);
```
Comment (3Hren) [inline src/context.cpp]: I think we should drop karma to the Hell here and elsewhere. But it's out of the scope of this PR.

File: src/context.cpp [hunk 5]
```diff
@@ -21,14 +21,18 @@
...
 #include "cocaine/detail/essentials.hpp"
-
+#include "cocaine/engine.hpp"
+#include "cocaine/format.hpp"
+#include "cocaine/idl/context.hpp"
 #include "cocaine/logging.hpp"
-
 #include "cocaine/rpc/actor.hpp"
 
+#include <boost/optional.hpp>
```
Comment (3Hren) [inline src/context.cpp]: It's too wide. Use `<boost/optional/optional.hpp>` instead.

File: include/cocaine/traits/tuple.hpp [hunk 6]
```diff
@@ -21,10 +21,10 @@
 #ifndef COCAINE_TYPELIST_SERIALIZATION_TRAITS_HPP
 #define COCAINE_TYPELIST_SERIALIZATION_TRAITS_HPP
 
-#include "cocaine/traits.hpp"
-
-#include "cocaine/platform.hpp"
+#include "cocaine/common.hpp"
+#include "cocaine/format.hpp"
```
Comment (3Hren) [inline include/cocaine/traits/tuple.hpp]: Why?

File: include/cocaine/context.hpp (hunk 1 above)
Comment (antmat) [inline include/cocaine/context.hpp]: I know, you have an allergy for "impl", but it's pimpl by design. context_data and other names will be misleading.

File: include/cocaine/traits/tuple.hpp (hunk 6 above)
Comment (antmat) [inline include/cocaine/traits/tuple.hpp]: line 134. And I can't forward template cocaine::format

File: src/context.cpp (hunk 5 above)
Comment (antmat) [inline src/context.cpp]: ok, I will, but it's cpp, I don't think it's a big problem here

File: src/context.cpp (hunk 3 above)
Comment (antmat) [inline src/context.cpp]: my ide is stupid, and we don't have clang-format for core :-(

File: src/service/logging.cpp (hunk 2 above)
Comment (antmat) [inline src/service/logging.cpp]: I doubt that someone will ever notice this one more call (which can be and probably will be inlined during lto) on the background of std::function call, but ok, I'll fix

**Comment by antmat** [comment]
updated

**Comment by 3Hren** [comment]
How about to make context an interface? This opens a door for unit testing.

**Comment by antmat** [comment]
What can we do with template method "invoke" in this case?

**Comment by antmat** [comment]
we can provide method for signal_hub retreival in this case. context_t::invoke only proxies to retroactive_signal::invoke

**Comment by 3Hren** [comment]
Good enough. Call it "notify".

Пт, 1 апр. 2016 г. в 11:19, Anton Matveenko <notifications@github.com>:

**Pull request merged**